packages/sanes-chrome-extension/src/logic/persona/persona.ts

```diff
@@ -0,0 +1,96 @@
+import { Amount } from '@iov/bcp';
+import { MultiChainSigner } from '@iov/core';
+
+import {
+  AccountManager,
+  AccountInfo,
+  createUserProfile,
+  AccountManagerChainConfig,
+} from '../user';
+import {
+  chainConnector,
+  getConfigurationFile,
+  codecTypeFromString,
+  algorithmForCodec,
+  pathBuilderForCodec,
+} from '../config';
+
+export class Persona {
+  /**
+   * Creates a new Persona instance.
+   *
+   * This function does everything that cannot be done in a constructor
+   * (because a constructor is synchonous): reading configs, connecting to the network,
+   * creating accounts.
+   */
+  public static async create(): Promise<Persona> {
+    const config = await getConfigurationFile();
+
+    const profile = await createUserProfile();
+    const signer = new MultiChainSigner(profile);
+
+    // connect chains
+    const managerChains: AccountManagerChainConfig[] = [];
+    for (const chainSpec of config.chains.map(chain => chain.chainSpec)) {
+      const codecType = codecTypeFromString(chainSpec.codecType);
+      const connector = chainConnector(codecType, chainSpec.bootstrapNodes);
+      const { connection } = await signer.addChain(connector);
+      managerChains.push({
+        chainId: connection.chainId(),
+        algorithm: algorithmForCodec(codecType),
+        derivePath: pathBuilderForCodec(codecType),
+      });
+    }
+
+    const manager = new AccountManager(profile, managerChains);
+
+    // Setup accounts
+    await manager.generateAccount(0);
+
+    return new Persona(signer, manager);
+  }
+
+  private readonly signer: MultiChainSigner;
+  private readonly accountManager: AccountManager;
+
+  /** The given signer and accountsManager must share the same UserProfile */
+  private constructor(
+    signer: MultiChainSigner,
+    accountManager: AccountManager
+  ) {
+    this.signer = signer;
+    this.accountManager = accountManager;
+  }
+
+  public mnemonic(): string {
+    return this.accountManager.mnemonic();
+  }
+
+  public async getAccounts(): Promise<ReadonlyArray<AccountInfo>> {
+    return this.accountManager.accounts();
+  }
+
+  public async getBalances(
+    accountIndex: number
+  ): Promise<ReadonlyArray<Amount>> {
+    const account = (await this.accountManager.accounts())[accountIndex];
+    const identities = [...account.identities.values()];
+    const pendingAccountResults = identities.map(identity => {
+      const { chainId, pubkey } = identity;
+      return this.signer.connection(chainId).getAccount({ pubkey });
+    });
+    const accountResults = await Promise.all(pendingAccountResults);
+
```

Comment:
Interesting, are you saying this:
```
const positiveResults = accountResults.filter(acc => !!acc);
```
does not work?

or even:
```
const balances = Array.from(accountResults.filter(acc => !!acc).map(acc => ...acc.balance));
```

Reply:
Right, because in `const positiveResults = accountResults.filter(acc1 => !!acc1).map(acc2 => ...acc2.balance)` the type of `acc2` is still `Account | undefined`. This can be avoided by writing an additional type checking method but unless we don't need this very often, I don't think it makes sense.